bots/sg_modify_scope_by_port.py

```diff
@@ -0,0 +1,195 @@
+        When '*' set for replacing any rule with the specific port
+# Examples:
+        AUTO: sg_modify_security_group_scope_by_port 22 0.0.0.0/0 10.0.0.0/24 inbound
+        AUTO: sg_modify_security_group_scope_by_port 22 * 10.0.0.0/24 inbound
+#Notes:
+    -  if the port is in a rule's port range the bot will change the rule's ip to desire ip , to avoid that
+      specify existing rule's scope instead of using '*'
+    - to split the rule around the port you can use the bot : #sg_single_rule_delete
+
+#Limitations: IPv6 is not supported yet
+
+"""
+import re
+
+PORT_TO = 'portTo'
+PORT_FROM = 'port'
+PROTOCOL = 'protocol'
+SCOPE = 'scope'
+ALL_TRAFFIC_PORT = 0
+ALL_TRAFFIC_PROTOCOL = '-1'
+
+"""
+checks if a rule exists in a security group , returns false/true 
+"""
+
+
+def rule_exists(sg, rule, direction, scope):
+    if direction.lower() == 'inbound':
+        for perm in sg.ip_permissions:
+            if perm['FromPort'] == rule[PORT_FROM] and perm['ToPort'] == rule[PORT_TO] and \
+                    perm['IpProtocol'] == rule[PROTOCOL].lower():
+                for ip in perm['IpRanges']:
+                    if ip['CidrIp'] == scope:  # found same rule exists
+                        return True;
+        return False
+    if direction.lower() == 'outbound':
+        for perm in sg.ip_permissions_egress:
+            if perm['FromPort'] == rule[PORT_FROM] and perm['ToPort'] == rule[PORT_TO] and \
+                    perm['IpProtocol'] == rule[PROTOCOL].lower():
+                for ip in perm['IpRanges']:
+                    if ip['CidrIp'] == scope:  # found same rule exists
+                        return True;
+        return False
+
+
+"""
+checks for ip validity , fix it to be right
+"""
+
+
+def fix_malformed(rule):
+    ip = re.split('/|\.', str(rule[SCOPE]))  # break ip to blocks
+    rule[SCOPE] = ip[0] + '.' + ip[1] + '.' + ip[2] + '.' + ip[3] + '/' + ip[4]
+    pass
+
+
+"""
+creates & removes the specified rules from a security group 
+"""
+
+
+def update_sg(sg, sg_id, rule, scope, direction, text_output):
+    # make sure that scope is in CIDR notation for example, 203.0.113.0/24
+    fix_malformed(rule)
+
+    if direction == 'inbound':
+        try:
+            sg.revoke_ingress(
+                CidrIp=rule[SCOPE],
+                FromPort=rule[PORT_FROM],
+                ToPort=rule[PORT_TO],
+                GroupId=sg_id,
+                IpProtocol=rule[PROTOCOL].lower()
+            )
+            text_output = text_output + ' rule : ' + rule[SCOPE] + ',' + str(rule[PORT_FROM]) + ',' + str(
+                rule[PORT_TO]) + ',' + \
+                          str(sg_id) + ',' + rule[PROTOCOL].lower() + ' deleted successfully ;'
+
+        except Exception as e:
+            text_output = text_output + f'Error while trying to delete security group. Error: {e}'
+
+        try:
+            # check if one is existing first ! avoid duplicates and exception
+            if rule_exists(sg, rule, direction, scope):
```

Comment:
Avoid using code inside if/else and create function with explicit name so the code will be readable

Reply:
code moved out of the if, and the name changed to is_rule_exists_in_sg